### PR TITLE
Make add operations on DefaultBlockDeletionContext thread safe

### PR DESCRIPTION
**Pull request by bradyoo**
Addresses https://github.com/Alluxio/alluxio/issues/12722

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java [hunk 1]
```diff
@@ -19,20 +19,21 @@
...
 import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * Class for tracking which blocks need to be deleted, and deleting them on close.
  */
 public final class DefaultBlockDeletionContext implements BlockDeletionContext {
   private final List<BlockDeletionListener> mListeners;
-  private final List<Long> mBlocks;
+  private final ConcurrentHashMap.KeySetView<Long, Boolean> mBlocks;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java:29]: Can we just use a concurrent list or queue? I think `ConcurrentLinkedQueue` might work?

**Review by gpang — COMMENTED** [review]
@bradyoo Thanks for this fix!

Can we also get a unit test? I think we could probably have a simple one where we force a parallel sync after we delete all the files in a directory in UFS. We should be able to get it to fail if there are multiple files and multiple blocks per file.

something like: 
- create ufs files (with multiple blocks) in a dir
- list dir (with sync)
- delete all those ufs files
- list dir (with sync)

File: core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java (hunk 1 above)
Comment (bradyoo) [reply core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java:29]: I don't think no matter how much we try to write this unit test, it's not really going to fail on the github actions machines.

**Comment by bradyoo** [comment]
I don't think no matter how much we try to write this unit test, it's not really going to fail on the github actions machines.

File: core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java (hunk 1 above)
Comment (bradyoo) [reply core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java:29]: I can do ConcurrentLinkedQueue. Usually though the ConcurrentHashMap related implementations tend to be the cleanest.

**Comment by bradyoo** [comment]
The fix you have requested is made though unit test is another thing.

**Review by gpang — APPROVED** [review]
LGTM

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Comment by bradyoo** [comment]
alluxio-bot, cherry-pick this to branch-2.4 please

**Comment by gpang** [comment]
alluxio-bot, cherry-pick this to branch-2.3 please

**Pull request merged**